github: add code coverage with codecov.io

arvindbr8 opened the pull request:
Fixes: #1676 

Codecov Reach graph - https://app.codecov.io/github/grpc/grpc-go/tree/code_coverage

RELEASE NOTES: none

dfawley commented on .github/workflows/coverage.yml:
```diff
@@ -0,0 +1,27 @@
+        append_coverage() {
+          [ -f $1 ] && (sed '1d' $1 >> ${GITHUB_WORKSPACE}/coverage.out)
+        }
+        go test -coverprofile=cov.out -covermode=set google.golang.org/grpc/...
```
I think to fix the cumulative coverage problem, we can just add `-coverpkg=./...` here?

dfawley commented on .github/workflows/coverage.yml:
```diff
@@ -0,0 +1,27 @@
+        }
+        go test -coverprofile=cov.out -covermode=set google.golang.org/grpc/...
+        append_coverage cov.out
+        for MOD_FILE in $(find . -name 'go.mod' | grep -Ev '^\./go\.mod'); do
```
Can we remove the `| grep -v` and not special case the root directory above?

arvindbr8 commented:
Discussed offline. I've made a few changes since the last push. 

We had a cumulative problem of `go test --cover` only covering unit tests. It did not include coverage from e2e style tests. For this adding `coverpkg=./...` worked which includes coverage of tests 

Since most of the changes are to the main module, it makes sense to limit the scope of presubmit check to only run `go test --cover` for the main module.

Ignored the following paths in codecov.
  - "test/"
  - "benchmark/"
  - "internal/testutils"
  - ".*.pb.go"
